core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,14 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
+    // yarn-standalone is deprecated, but still supported
+    if ((master == "yarn-cluster" || master == "yarn-standalone") &&
+        !_conf.contains("spark.yarn.app.id")) {
```

Comment:
This is an early and clean solution, rather than checking for the master. YARN cluster mode the SparkContext is launched as a separate user thread at container 0 and joined later.

@srowen If you resolve it, I will remove `yarn` from the configuration parameter, but you can do it afterwards. Just to keep pull-request timeline.